iceoryx_dust/vocabulary/include/iox/span.hpp

```diff
@@ -0,0 +1,451 @@
+#include "iox/type_traits.hpp"
+#include "iox/uninitialized_array.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <cstdio>
+#include <limits>
+
+using namespace std;
+
+namespace iox
+{
+// constants
+constexpr uint64_t DYNAMIC_EXTENT = std::numeric_limits<uint64_t>::max();
+
+// Implementation of C++17 std::size() and std::data()
+
+/// @brief Returns c.size(), converted to the return type if necessary.
+/// @tparam C Type of the container
+/// @param c A container or view with a size member function
+/// @return The size of c
+template <typename Container>
+constexpr auto size(const Container& container) -> decltype(container.size());
+
+/// @brief Returns N
+/// @tparam T Type of the array
+/// @tparam N Size of the array
+/// @param An array
+/// @return Size of the array
+template <typename T, std::uint64_t N>
+constexpr std::uint64_t size(const T (&)[N]) noexcept;
+
+/// @brief Returns N
+/// @tparam T Type of the iox::UninitializedArray
+/// @tparam N Size of the iox::UninitializedArray
+/// @param An iox::UninitializedArray
+/// @return Size of the iox::UninitializedArray
+template <typename T, std::uint64_t N, template <typename, uint64_t> class Buffer>
+constexpr std::uint64_t size(const UninitializedArray<T, N, Buffer>&) noexcept;
+
+/// @brief Returns a pointer to the block of memory containing the elements of the range.
+/// @tparam C Type of the container
+/// @param c A container or view with a data() member function
+/// @return Returns c.data()
+template <typename Container>
+constexpr auto data(Container& container) -> decltype(container.data());
+
+/// @brief Returns a pointer to the block of memory containing the elements of the range.
+/// @tparam C Type of the container
+/// @param c A container or view with a data() member function
+/// @return Returns c.data()
+template <typename Container>
+constexpr auto data(const Container& container) -> decltype(container.data());
+
+/// @brief Returns a pointer to the block of memory containing the elements of the range.
+/// @tparam T Type of the array
+/// @tparam N Size of the array
+/// @param array An array of arbitrary type
+/// @return Returns array
+template <typename T, std::uint64_t N>
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+constexpr T* data(T (&array)[N]) noexcept;
+
+/// @brief Returns a pointer to the block of memory containing the elements of the range.
+/// @tparam T Type of the iox::UninitializedArray
+/// @tparam N Size of the iox::UninitializedArray
+/// @param array An iox::UninitializedArray of arbitrary type
+/// @return Returns iox::UninitializedArray
+template <typename T, std::uint64_t N, template <typename, uint64_t> class Buffer>
+constexpr T* data(UninitializedArray<T, N, Buffer>& uninit_array) noexcept;
+
+template <typename T, uint64_t Extent = DYNAMIC_EXTENT>
+class span;
+
+namespace detail
+{
+template <uint64_t I>
+using size_constant = std::integral_constant<uint64_t, I>;
```

Comment:
I'm wondering if it would make sense to move this to a more central location. `size_constant` is a quite generic name and I fear some issues regarding double definition ... not like compile errors but having this in multiple translation units and then getting some weird linker issues

Reply:
Renamed.